agent/cache-types/streaming.go

```diff
@@ -2,11 +2,174 @@ package cachetype
 
 import (
 	"context"
+	"sync"
+	"time"
 
+	"github.com/hashicorp/consul/agent/cache"
 	"github.com/hashicorp/consul/agent/consul/stream"
 	"google.golang.org/grpc"
 )
 
-type GRPCClient interface {
+type StreamingClient interface {
 	Subscribe(ctx context.Context, in *stream.SubscribeRequest, opts ...grpc.CallOption) (stream.Consul_SubscribeClient, error)
 }
+
+type EventHandler interface {
```

Comment:
Maybe document that HandleEvent and State are called from different goroutines and so should syncronise access to any internal state they need.